### PR TITLE
ARROW-6847: [C++] Add range_expression adapter to Iterator

### PR DESCRIPTION
**Pull request by bkietz**
Allows `Iterator<T>` to be used in a range-for loop

### CODE REVIEW

**Comment by pitrou** [comment]
Getting the Status on the side is a bit weird. Should we return a `Result<>` when iterating from the range instead?
@JohanMabille @wolfv What would be your thoughts on this?

**Comment by pitrou** [comment]
(also needs rebasing @bkietz )

**Comment by bkietz** [comment]
@pitrou returning a `Result<>` as the range value requires placing a status check inside the loop body:

```c++
for (Result<int> i_result : ints.AsRange()) {
  ARROW_ASSIGN_OR_RAISE(int i, i_result);
  // loop body
}
```

IMHO, this seems less natural to me than placing the check outside:

```c++
Status status;
for (int i : ints.AsRange(&status)) {
  // loop body
}
RETURN_NOT_OK(status);
```

But neither is a problem. @fsaintjacques ?

**Comment by bkietz** [comment]
@fsaintjacques @pitrou I've refactored so the range's elements are `Result<>`

File: cpp/src/arrow/result.h [hunk 1]
```diff
@@ -215,6 +215,9 @@ class Result {
     return *this;
   }
 
+  /// Compare to another Result.
+  bool operator==(const Result& other) const { return variant_ == other.variant_; }
```
Comment (pitrou) [inline cpp/src/arrow/result.h]: Does C++ create the corresponding `operator!=` automatically?

**Review by pitrou — COMMENTED** [review]
Perhaps you should add some tests for Status and Result (in)equality?

File: cpp/src/arrow/result.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/result.h]: No, I'll add it

**Comment by bkietz** [comment]
@pitrou I've added inequality and tests

@fsaintjacques The `EqualityComparable` helper ought to help with https://issues.apache.org/jira/browse/ARROW-6772

**Comment by JohanMabille** [comment]
> Getting the Status on the side is a bit weird. Should we return a Result<> when iterating from the range instead?
@JohanMabille @wolfv What would be your thoughts on this?

Not sure to get why you need the iterator to fill a Status object. Anyway, the second loop written by @bkietz is indeed more natural in C++ than having to check anything in the loop body.

**Comment by bkietz** [comment]
@JohanMabille the Iterator may return an error status when attempting to retrieve the next sequence element. That error needs to be accessible by the consumer of the Iterator, hence the result/side status loops.

File: cpp/src/arrow/util/compare.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace arrow {
+namespace util {
+
+/// CRTP helper for declaring equality comparison. Defines operator== and operator!=
+template <typename T>
+class EqualityComparable {
+ public:
+  ~EqualityComparable() {
+    static_assert(
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/compare.h]: Is this safe to override the destructor just for a static_assert?

File: cpp/src/arrow/util/compare.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include <utility>
+
+#include "arrow/util/macros.h"
+
+namespace arrow {
+namespace util {
+
+/// CRTP helper for declaring equality comparison. Defines operator== and operator!=
+template <typename T>
+class EqualityComparable {
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/compare.h]: Nice utility. Bonus point if you make one for printable(ToString), we'll gain gtest pretty printters.

File: cpp/src/arrow/util/iterator.h [hunk 4]
```diff
@@ -88,10 +90,51 @@ class Iterator {
   }
 
   /// Iterators will only compare equal if they are both null
-  bool operator==(const Iterator& other) const { return ptr_ == other.ptr_; }
+  bool Equals(const Iterator& other) const { return ptr_ == other.ptr_; }
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/iterator.h]: Could you add explanation that this is required to check the `end()` condition? It wasn't obvious when I first looked.

**Review by fsaintjacques — APPROVED** [review]
Very handy work, this will definitively improve the ergonomics. I've restarted the tests, failed due to external errors (github or dockerhub network).

File: cpp/src/arrow/util/compare.h (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/util/compare.h]: The default generated destructor is just an empty block like this. I think it's fine, but if you prefer I can replace with `void CheckEqualsImpemented()` or so

File: cpp/src/arrow/util/iterator.h (hunk 4 above)
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: Will do

File: cpp/src/arrow/util/compare.h (hunk 3 above)
Comment (bkietz) [reply cpp/src/arrow/util/compare.h]: I'll do that

**Pull request closed without merging**